pkg/route/secret/secret_monitor.go

```diff
@@ -0,0 +1,215 @@
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+type SecretEventHandlerRegistration interface {
+	cache.ResourceEventHandlerRegistration
+
+	GetKey() ObjectKey
+	GetHandler() cache.ResourceEventHandlerRegistration
+}
+
+type SecretMonitor interface {
+	//
+	AddSecretEventHandler(namespace, routeSecretName string, handler cache.ResourceEventHandler) (SecretEventHandlerRegistration, error)
+	//
+	RemoveSecretEventHandler(SecretEventHandlerRegistration) error
+	//
+	GetSecret(SecretEventHandlerRegistration) (*v1.Secret, error)
+}
+
+type secretEventHandlerRegistration struct {
+	cache.ResourceEventHandlerRegistration
+	// objectKey will populate during AddEventHandler, and will be used during RemoveEventHandler
+	objectKey ObjectKey
+}
+
+func (r *secretEventHandlerRegistration) GetKey() ObjectKey {
+	return r.objectKey
+}
+
+func (r *secretEventHandlerRegistration) GetHandler() cache.ResourceEventHandlerRegistration {
+	return r.ResourceEventHandlerRegistration
+}
+
+type secretMonitor struct {
+	kubeClient kubernetes.Interface
+
+	lock sync.RWMutex
+	// monitors is map of singleItemMonitor. Each singleItemMonitor monitors/watches
+	// a secret through individual informer.
+	monitors map[ObjectKey]*singleItemMonitor
+}
+
+func NewSecretMonitor(kubeClient kubernetes.Interface) SecretMonitor {
+	return &secretMonitor{
+		kubeClient: kubeClient,
+		monitors:   map[ObjectKey]*singleItemMonitor{},
+	}
+}
+
+// create/update secret watch.
+// routeSecretName is a combination of "routename_secretname"
+func (s *secretMonitor) AddSecretEventHandler(namespace, routeSecretName string, handler cache.ResourceEventHandler) (SecretEventHandlerRegistration, error) {
+	return s.addSecretEventHandler(namespace, routeSecretName, handler, nil)
+}
+
+// addSecretEventHandler should only be used directly for tests. For production use AddSecretEventHandler().
+// createInformerFn helps in mocking sharedInformer for unit tests.
+func (s *secretMonitor) addSecretEventHandler(namespace, routeSecretName string, handler cache.ResourceEventHandler, createInformerFn func() cache.SharedInformer) (SecretEventHandlerRegistration, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if !validateName(routeSecretName) {
+		return nil, fmt.Errorf("invalid routeSecretName combination : %s", routeSecretName)
+	}
+
+	// Inside a namespace multiple route can refer a common secret.
+	// Hence route1_secret and route2_secret should denote separate key,
+	// so separate monitors inside a namespace.
+	key := NewObjectKey(namespace, routeSecretName)
+
+	// Check if monitor / watch already exists
+	m, exists := s.monitors[key]
+	if !exists {
+		secretName, err := getSecretName(routeSecretName, true)
+		if err != nil {
+			return nil, err
+		}
+
+		var sharedInformer cache.SharedInformer
+		if createInformerFn == nil {
```

Comment:
set your desired function to a default value of `createInformer` and inject as the exception to avoid the nesting here.

Reply:
I think the default value would be `nil` for general flow. This nesting is required for unit tests only to inject a fake sharedInformer.  I tried thinking to do it differently, but did not find better way. Can you please suggest if there is another way?

Reply:
Another way could be the following, let me know if this looks better

```
diff --git a/pkg/route/secret/secret_monitor.go b/pkg/route/secret/secret_monitor.go
index 617a5666c..a06e5c3bd 100644
--- a/pkg/route/secret/secret_monitor.go
+++ b/pkg/route/secret/secret_monitor.go
@@ -61,7 +61,22 @@ func NewSecretMonitor(kubeClient kubernetes.Interface) SecretMonitor {
 
 // create secret watch.
 func (s *secretMonitor) AddSecretEventHandler(namespace, secretName string, handler cache.ResourceEventHandler) (SecretEventHandlerRegistration, error) {
-	return s.addSecretEventHandler(namespace, secretName, handler, nil)
+	return s.addSecretEventHandler(namespace, secretName, handler, s.createSecretInformer(namespace, secretName))
+}
+
+func (s *secretMonitor) createSecretInformer(namespace, name string) func() cache.SharedInformer {
+	return func() cache.SharedInformer {
+		return cache.NewSharedInformer(
+			cache.NewListWatchFromClient(
+				s.kubeClient.CoreV1().RESTClient(),
+				"secrets",
+				namespace,
+				fields.OneTermEqualSelector("metadata.name", name),
+			),
+			&corev1.Secret{},
+			0,
+		)
+	}
 }
 
 // addSecretEventHandler should only be used directly for tests. For production use AddSecretEventHandler().
@@ -82,24 +97,7 @@ func (s *secretMonitor) addSecretEventHandler(namespace, secretName string, hand
 
 	// start secret informer
 	if !exists {
-		var sharedInformer cache.SharedInformer
-		if createInformerFn == nil {
-			// create a single secret monitor
-			sharedInformer = cache.NewSharedInformer(
-				cache.NewListWatchFromClient(
-					s.kubeClient.CoreV1().RESTClient(),
-					"secrets",
-					namespace,
-					fields.OneTermEqualSelector("metadata.name", secretName),
-				),
-				&corev1.Secret{},
-				0,
-			)
-		} else {
-			// only for testability
-			klog.Warning("creating informer for testability")
-			sharedInformer = createInformerFn()
-		}
+		sharedInformer := createInformerFn()
 
 		m = newSingleItemMonitor(key, sharedInformer)
 		go m.StartInformer()
```